### PR TITLE
feat: express route to sign payloads in external signing mode

### PR DESCRIPTION
**Pull request by manav-c**
## Description

This route will enable our Clients and Partners to leverage BitGo Express in External Signing mode to sign arbitrary http post request payloads, with their trading account key, in order to integrate with our systems.

## Issue Number

Ticket: BG-74347. 

## Type of change

Please delete options that are not relevant.

- [X] New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

### Manual testing using TEST data:
I stood up two local instances of the express server; one in external signing mode and the other not. 
My external signing mode server was running on port 4000. Whereas the other one (not in ext signing mode) was on port 3000. 
On the server running on 3000:
1. Set `BITGO_DISABLE_SSL=true` in `.env`
2. Set `BITGO_EXTERNAL_SIGNER_URL=localhost:4000` in `.env`
3. Start the server

Server running on 4000: 
1. Choose a trading account wallet from one of your enterprises. Copy its walletID. This is the trading account whose key will be used to sign the payload. 
5. Populate `BITGO_EXTERNAL_SIGNER_ACCESS_TOKEN` and `BITGO_EXTERNAL_SIGNER_WALLET_IDS` env vars.
6. Use `fetchEncryptedKeys.ts` to fetch the encrypted private key for that walletID to update the `encryptedPrivKeys.json` file. 
7. Set `BITGO_SIGNER_MODE=true` 
8. Set `BITGO_SIGNER_FILE_SYSTEM_PATH` to the path to the `encryptedPrivKeys.json` file. 
9. Start the server

In postman: 

POST `http://localhost:3000/api/v2/ofc/signPayload` with the following payload: `payload` can be any arbitrary payload. 
```
{
    "walletId": "<WALLET_ID chosen above>",
    "payload": {
        "this": {
            "is": {
                "a": "payload"
            }
        }
    }
}
```

### Manual E2E testing in DEV:

- Created an enterprise along with its trading account.
- Added the walletId (trading account ID) and other required info (local encrypted keys in a json + passphrase for that walletID) on to a local express server set up in external signing mode (port 4000).
- Started up a local express server not in ext signing mode (port 3000) with externalSigningURL set to the server running on port 4000. 
- Signed an arbitrary payload for the walletID using the express server that is running on port 3000. This server proxies the request to localhost:4000. 
- Added the payload, signature on to DAS entries in develop by port-forwarding DAS and using the `POST /entries` endpoint. 

# Checklist:

- [X] My code follows the style guidelines of this project
- [X] I have performed a self-review of my own code
- [x] My code compiles correctly for both Node and Browser environments
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] My commits follow [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/) and I have properly described any BREAKING CHANGES
- [x] The ticket or github issue was included in the commit message as a reference
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: modules/express/src/clientRoutes.ts [hunk 1]
```diff
@@ -461,11 +461,76 @@ export async function handleV2Sign(req: express.Request) {
   }
 }
 
+export async function handleV2OFCSignPayloadInExtSigningMode(
+  req: express.Request
+): Promise<{ payload: string; signature: string }> {
+  const walletId = req.body.walletId;
+  const payload = req.body.payload;
```
Comment (zahin-mohammad) [inline modules/express/src/clientRoutes.ts:468]: Nit: we can define a codec for this request and use io-ts.

File: modules/express/test/unit/clientRoutes/signPayload.ts [hunk 2]
```diff
@@ -69,3 +68,63 @@ describe('Sign an arbitrary payload with trading account key', function () {
...
+        a: 'payload',
+      },
+    },
+  };
+
+  before(() => {
+    bitgo = new BitGo({ env: 'test' });
+  });
+
+  it('should work', async () => {
```
Comment (zahin-mohammad) [inline modules/express/test/unit/clientRoutes/signPayload.ts:94]: Please have a more descriptive test name.

File: modules/express/test/unit/clientRoutes/signPayload.ts [hunk 3]
```diff
@@ -69,3 +68,63 @@ describe('Sign an arbitrary payload with trading account key', function () {
     await handleV2OFCSignPayload(req).should.be.resolvedWith(expectedResponse);
   });
 });
+
+describe('Sign an arbitrary payload with trading account key read from the local file system (external signing mode)', () => {
```
Comment (zahin-mohammad) [inline modules/express/test/unit/clientRoutes/signPayload.ts:72]: Please add some non happy path tests.

**Review by zahin-mohammad — CHANGES_REQUESTED** [review]

File: modules/express/test/unit/clientRoutes/signPayload.ts (hunk 2 above)
Comment (manav-c) [reply modules/express/test/unit/clientRoutes/signPayload.ts:94]: ah I meant to update this. thanks!

File: modules/express/test/unit/clientRoutes/signPayload.ts (hunk 3 above)
Comment (manav-c) [reply modules/express/test/unit/clientRoutes/signPayload.ts:73]: updated!

File: modules/express/src/clientRoutes.ts (hunk 1 above)
Comment (manav-c) [reply modules/express/src/clientRoutes.ts:468]: Thanks for the suggestion! This feature is currently blocking testing with a partner and so I want to keep the scope for this ticket as low as possible and get it over the finish line ASAP. Will keep this in mind for a future iteration on this feature.

File: modules/express/test/unit/clientRoutes/signPayload.ts [hunk 4]
```diff
@@ -69,3 +69,158 @@ describe('Sign an arbitrary payload with trading account key', function () {
...
+      })
+    );
+    readFileStub.restore();
+    envStub.restore();
+  });
+
+  describe('with invalid setup', () => {
+    const invalidPrv = '{"61f039aad587c2000745c687373e0fa9":"invalid"}';
+  
+    it('with missing wallet passphrase in env, should throw error', async () => {
```
Comment (zahin-mohammad) [inline modules/express/test/unit/clientRoutes/signPayload.ts:137]: nit, the `describe` title + `it` title should read like a sentence. 

one example: 
`with an invalid setup, it should throw an error due to an missing wallet passphrase in the env`

File: modules/express/test/unit/clientRoutes/signPayload.ts [hunk 5]
```diff
@@ -69,3 +69,158 @@ describe('Sign an arbitrary payload with trading account key', function () {
...
+          walletId,
+          payload,
+        },
+      } as unknown as Request;
+
+      try {
+        await handleV2OFCSignPayloadInExtSigningMode(req);
+      } catch (err) {
+        err.message.should.containEql('Could not find wallet passphrase');
+      }
```
Comment (zahin-mohammad) [inline modules/express/test/unit/clientRoutes/signPayload.ts:150]: This test actually has a subtle flaw. If the function never throws an error, then this tests would still pass. 
One way to fix it is to manually throw in the try block.

Or we can use `should.js` to do this for us.


```suggestion
     await handleV2OFCSignPayloadInExtSigningMode(req).rejectedWith(`Could not find wallet passphrase`
```

Same comment to all the other non happy path tests.

**Review by zahin-mohammad — CHANGES_REQUESTED** [review]

File: modules/express/test/unit/clientRoutes/signPayload.ts (hunk 5 above)
Comment (manav-c) [reply modules/express/test/unit/clientRoutes/signPayload.ts:150]: updated, thanks!

File: modules/express/test/unit/clientRoutes/signPayload.ts [hunk 6]
```diff
@@ -69,3 +69,144 @@ describe('Sign an arbitrary payload with trading account key', function () {
...
+
+    await handleV2OFCSignPayloadInExtSigningMode(req).should.be.resolvedWith(expectedResponse);
+    readFileStub.should.be.calledOnceWith('signerFileSystemPath');
+    signMessageStub.should.be.calledOnceWith(
+      sinon.match({
+        prv: secret,
+      })
+    );
+    readFileStub.restore();
+    envStub.restore();
```
Comment (zahin-mohammad) [inline modules/express/test/unit/clientRoutes/signPayload.ts:131]: Nit: can be done in an afterEach block.

File: modules/express/test/unit/clientRoutes/signPayload.ts [hunk 7]
```diff
@@ -69,3 +69,144 @@ describe('Sign an arbitrary payload with trading account key', function () {
...
+      is: {
+        a: 'payload',
+      },
+    },
+  };
+
+  before(() => {
+    bitgo = new BitGo({ env: 'test' });
+  });
+
```
Comment (zahin-mohammad) [inline modules/express/test/unit/clientRoutes/signPayload.ts:94]: Should prob add a sinon.restore in an afterEach block here.

**Review by zahin-mohammad — APPROVED** [review]

**Comment by jonball4** [comment]
👍

**Pull request merged**